### PR TITLE
Fix CreateObject support for JsonConstructor types

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This should make tests passing again.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs [hunk 1]
```diff
@@ -23,6 +23,14 @@ internal abstract partial class ObjectWithParameterizedConstructorConverter<T> :
 
         internal sealed override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, [MaybeNullWhen(false)] out T value)
         {
+            JsonTypeInfo jsonTypeInfo = state.Current.JsonTypeInfo;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:26]: I think you might need similar fix in LargeObjectWithParametrizedConsturctor (or whatever it's called) - you might want to copy&paste the tests and add at least 4-5 args to hit that code path

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs [hunk 2]
```diff
@@ -40,7 +40,12 @@ internal ReflectionJsonTypeInfo(JsonConverter converter, JsonSerializerOptions o
                 AddPropertiesAndParametersUsingReflection();
             }
 
-            SetCreateObject(Options.MemberAccessorStrategy.CreateConstructor(typeof(T)), useForExtensionDataProperty: true);
+            Func<object>? createObject = Options.MemberAccessorStrategy.CreateConstructor(typeof(T));
+            if (converter.UsesDefaultConstructor)
+            {
+                SetCreateObject(createObject);
+            }
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:47]: nit: enter after closing brace

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs [hunk 3]
```diff
@@ -130,6 +130,19 @@ internal static bool ShouldFlush(Utf8JsonWriter writer, ref WriteStack state)
...
         internal virtual bool ConstructorIsParameterized { get; }
 
+        /// <summary>
+        ///  For reflection-based metadata generation, indicates whether the
+        ///  converter avails of default constructors when deserializing types.
+        /// </summary>
+        internal bool UsesDefaultConstructor =>
+            ConverterStrategy switch
+            {
+                ConverterStrategy.Object => !ConstructorIsParameterized && this is not ObjectConverter,
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs:140]: why `not ObjectConverter`?

Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs:140]: if this is converter for `typeof(object)` then disregard the question 😄

**Review by krwq — APPROVED** [review]
LGTM assuming tests are passing, double check LargeObjectWithParametrizedCtor

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs:140]: Yep that's the one!

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:26]: It's deriving from that class, but you're right that we should add another test for that.

**Comment by krwq** [comment]
thanks!

**Pull request merged**